### PR TITLE
[ENG-7739][steps] add support for running js code in functions

### PR DESCRIPTION
**Pull request by dsokal**
# Why

Currently, steps and functions use only bash scripts. For implementing functions that share the JS context with the runner, we need to be able to run any arbitrary JS code. This'll let us implement features like artifact upload without hacks. The artifact upload logic (uploading to a particular storage like S3/GCS) shouldn't be part of the library. Instead, it should be injected when creating the workflow object. With the current approach, where there's a bash command upload-artifact  that can be run from the user's script, we can't really share the upload implementation with that script.

# How

- https://github.com/expo/eas-build/pull/201/commits/272edc85a8b0ab9c6a19637691c872a8bcec6ce7 removes the build context object from the `BuildFunction` constructor. `BuildFunction`s are going to be used for defining EAS built-in functions. The context is not available at that time.
- https://github.com/expo/eas-build/pull/201/commits/386edc41f010190d1f45c96c9f16638dd15aad54 renames input / output creators to providers. I wanted to make a function for creating creators but `createInputCreator` sounded a bit funny.
- https://github.com/expo/eas-build/pull/201/commits/4996a465d47ae12e067f8eeb96f92937c8e7ebda introduces the `.createProvider` static methods to `BuildStepInput` and `BuildStepOutput` to streamline provider creation.
- https://github.com/expo/eas-build/pull/201/commits/54c87469a3cd33c4f5f917322f2b692755ed7cc2 changes `BuildStep` to either run the `command` (bash script) or JS function provided to the constructor.
- https://github.com/expo/eas-build/pull/201/commits/c9cbc851fe8a583d9bbef8ace0d1d6dbffcf6129 fixes setting IDs for function and step objects.
- https://github.com/expo/eas-build/pull/201/commits/e30bc631c8175c8cc16ac7694a233d7fe9a13a56 adds namespaces for functions so we can define functions like `eas/upload_artifacts`.
- https://github.com/expo/eas-build/pull/201/commits/dc4ea8a8d849999a03378e96bb3ababb75f398aa makes `BuildConfigParser` accept external functions for building workflow object.

# Test Plan

Tests.

### CODE REVIEW

File: packages/steps/src/utils/expodash/duplicates.ts [hunk 1]
```diff
@@ -0,0 +1,19 @@
+import { uniq } from './uniq.js';
+
+export function duplicates<T = any>(items: T[]): T[] {
+  const uniqueItems = uniq(items);
+  if (uniqueItems.length === items.length) {
+    return [];
+  }
+
```
Comment (wkozyra95) [inline packages/steps/src/utils/expodash/duplicates.ts:8]:
```suggestion
```

An early exit would normally have performance befits, but most likely calling uniq is doing a similar amount of work as the rest of this function. Other than performance there is no benefit in keeping early exit code, it only complicates code by adding additional cases that regular implementation already handles correctly.

File: packages/steps/src/BuildStep.ts [hunk 2]
```diff
@@ -224,11 +260,11 @@ export class BuildStep {
     };
   }
 
-  private getStepDisplayName(name: string | undefined, command: string): string | undefined {
+  private getStepDisplayName(name: string | undefined, command?: string): string | undefined {
     if (name) {
       return name;
     }
-    if (command !== '') {
+    if (command !== undefined && command !== '') {
```
Comment (wkozyra95) [inline packages/steps/src/BuildStep.ts:267]: if you checking for both empty string and undefined why not just `!command`?

File: packages/steps/src/BuildFunction.ts [hunk 3]
```diff
@@ -1,78 +1,96 @@
...
-    shell,
-  }: {
-    id?: string;
-    callInputs?: BuildFunctionCallInputs;
-    workingDirectory: string;
-    shell?: string;
-  }): BuildStep {
-    const buildStepId = id ?? this.id ?? uuidv4();
+  public getFullId(): string {
+    return this.namespace === undefined ? this.id : `${this.namespace}/${this.id}`;
```
Comment (wkozyra95) [inline packages/steps/src/BuildFunction.ts:61]: If we are using `/` as namespace separator we should limit what can be used for regular ids. I would suggest limiting it to `^(?:eas\/)?[a-zA-Z0-9_-]*$` (not tested).

**Review by wkozyra95 — APPROVED** [review]

**Review by szdziedzic — APPROVED** [review]
LGTM, although tests are failing: https://github.com/expo/eas-build/actions/runs/4447772532/jobs/7809714154?pr=201 (seems like minor issue)

**Pull request merged**